File: typescript/tsfind/src/fileresultsorter.ts
/*
 * fileresultsorter.ts
 *
 * FileResultSorter class provides sorting of search results
 */

import {FileResult} from './fileresult';
import {FindSettings} from './findsettings';
import {SortBy} from "./sortby";

"use strict";

export class FileResultSorter {
    settings: FindSettings;

    constructor(settings: FindSettings) {
        this.settings = settings;
    }

    public cmpFileResultsByPath(fr1: FileResult, fr2: FileResult): number {
        const [path1, path2]: string[] = this.settings.sortCaseInsensitive ?
            [fr1.path.toLowerCase(), fr2.path.toLowerCase()] :
            [fr1.path, fr2.path];
        if (path1 === path2) {
            const [filename1, filename2]: string[] = this.settings.sortCaseInsensitive ?
                [fr1.fileName.toLowerCase(), fr2.fileName.toLowerCase()] :
                [fr1.fileName, fr2.fileName];
            if (filename1 === filename2) return 0;
            return filename1 < filename2 ? -1 : 1;
        }
        return path1 < path2 ? -1 : 1;
    }

    public cmpFileResultsByName(fr1: FileResult, fr2: FileResult): number {
        const [fileName1, fileName2]: string[] = this.settings.sortCaseInsensitive ?
            [fr1.fileName.toLowerCase(), fr2.fileName.toLowerCase()] :
            [fr1.fileName, fr2.fileName];
        if (fileName1 === fileName2) {
            const [path1, path2]: string[] = this.settings.sortCaseInsensitive ?
                [fr1.path.toLowerCase(), fr2.path.toLowerCase()] :
                [fr1.path, fr2.path];
            if (path1 === path2) return 0;
            return path1 < path2 ? -1 : 1;
        }
        return fileName1 < fileName2 ? -1 : 1;
    }

    public cmpFileResultsBySize(fr1: FileResult, fr2: FileResult): number {
        if (fr1.fileSize === fr2.fileSize) {
            return this.cmpFileResultsByPath(fr1, fr2);
        }
        return fr1.fileSize < fr2.fileSize ? -1 : 1;
    }

    public cmpFileResultsByType(fr1: FileResult, fr2: FileResult): number {
        if (fr1.fileType === fr2.fileType) {
            return this.cmpFileResultsByPath(fr1, fr2);
        }
        return fr1.fileType < fr2.fileType ? -1 : 1;
    }

    public cmpFileResultsByLastMod(fr1: FileResult, fr2: FileResult): number {
        if (fr1.lastMod === fr2.lastMod) {
            return this.cmpFileResultsByPath(fr1, fr2);
        }
        return fr1.lastMod < fr2.lastMod ? -1 : 1;
    }

    public getFileResultComparator(): (a: FileResult, b: FileResult) => number {
        if (this.settings.sortDescending) {
            switch (this.settings.sortBy) {
                case SortBy.FileName:
                    return (a, b) => this.cmpFileResultsByName(b, a);
                case SortBy.FileSize:
                    return (a, b) => this.cmpFileResultsBySize(b, a);
                case SortBy.FileType:
                    return (a, b) => this.cmpFileResultsByType(b, a);
                case SortBy.LastMod:
                    return (a, b) => this.cmpFileResultsByLastMod(b, a);
                default:
                    return (a, b) => this.cmpFileResultsByPath(b, a);
            }
        }
        switch (this.settings.sortBy) {
            case SortBy.FileName:
                return (a, b) => this.cmpFileResultsByName(a, b);
            case SortBy.FileSize:
                return (a, b) => this.cmpFileResultsBySize(a, b);
            case SortBy.FileType:
                return (a, b) => this.cmpFileResultsByType(a, b);
            case SortBy.LastMod:
                return (a, b) => this.cmpFileResultsByLastMod(a, b);
            default:
                return (a, b) => this.cmpFileResultsByPath(a, b);
        }
    }

    public sort(fileResults: FileResult[]): void {
        const sortComparator = this.getFileResultComparator();
        fileResults.sort(sortComparator);
    }
}

File: cpp/cppfind/src/FileResultSorter.cpp
#include "FileResultSorter.h"

namespace cppfind {
    FileResultSorter::FileResultSorter(const FindSettings& settings) :  m_settings{settings} {
    }

    FileResultSorter::FileResultSorter(const std::unique_ptr<FindSettings>& settings_ptr) : m_settings{*settings_ptr} {
    }

    FindSettings FileResultSorter::settings() const {
        return m_settings;
    }

    int cmp_file_results_by_path(const FileResult& fr1, const FileResult& fr2) {
        if (const int path_cmp = fr1.file_path().parent_path().compare(fr2.file_path().parent_path()); path_cmp != 0) {
            return path_cmp;
        }
        return fr1.file_path().filename().compare(fr2.file_path().filename());
    }

    int cmp_file_results_by_path_ci(const FileResult& fr1, const FileResult& fr2) {
        if (const int path_cmp = strcasecmp(fr1.file_path().parent_path().c_str(), fr2.file_path().parent_path().c_str()); path_cmp != 0) {
            return path_cmp;
        }
        return strcasecmp(fr1.file_path().filename().c_str(), fr2.file_path().filename().c_str());
    }

    std::function<bool(FileResult&, FileResult&)> FileResultSorter::get_cmp_file_results_by_path() const {
        if (m_settings.sort_descending()) {
            if (m_settings.sort_case_insensitive()) {
                return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_path_ci(fr2, fr1) <= 0; };
            }
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_path(fr2, fr1) <= 0; };
        }
        if (m_settings.sort_case_insensitive()) {
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_path_ci(fr1, fr2) <= 0; };
        }
        return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_path(fr1, fr2) <= 0; };
    }

    int cmp_file_results_by_name(const FileResult& fr1, const FileResult& fr2) {
        if (const int file_cmp = fr1.file_path().filename().compare(fr2.file_path().filename()); file_cmp != 0) {
            return file_cmp;
        }
        return fr1.file_path().parent_path().compare(fr2.file_path().parent_path());
    }

    int cmp_file_results_by_name_ci(const FileResult& fr1, const FileResult& fr2) {
        if (const int file_cmp = strcasecmp(fr1.file_path().filename().c_str(), fr2.file_path().filename().c_str()); file_cmp != 0) {
            return file_cmp < 0;
        }
        return strcasecmp(fr1.file_path().parent_path().c_str(), fr2.file_path().parent_path().c_str()) < 0;
    }

    std::function<bool(FileResult&, FileResult&)> FileResultSorter::get_cmp_file_results_by_name() const {
        if (m_settings.sort_descending()) {
            if (m_settings.sort_case_insensitive()) {
                return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_name_ci(fr2, fr1) <= 0; };
            }
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_name(fr2, fr1) <= 0; };
        }
        if (m_settings.sort_case_insensitive()) {
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_name_ci(fr1, fr2) <= 0; };
        }
        return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_name(fr1, fr2) <= 0; };
    }

    int cmp_file_results_by_size(const FileResult& fr1, const FileResult& fr2) {
        if (fr1.file_size() == fr2.file_size()) {
            return cmp_file_results_by_path(fr1, fr2);
        }
        return fr1.file_size() < fr2.file_size() ? -1 : 1;
    }

    int cmp_file_results_by_size_ci(const FileResult& fr1, const FileResult& fr2) {
        if (fr1.file_size() == fr2.file_size()) {
            return cmp_file_results_by_path_ci(fr1, fr2);
        }
        return fr1.file_size() < fr2.file_size() ? -1 : 1;
    }

    std::function<bool(FileResult&, FileResult&)> FileResultSorter::get_cmp_file_results_by_size() const {
        if (m_settings.sort_descending()) {
            if (m_settings.sort_case_insensitive()) {
                return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_size_ci(fr2, fr1) <= 0; };
            }
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_size(fr2, fr1) <= 0; };
        }
        if (m_settings.sort_case_insensitive()) {
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_size_ci(fr1, fr2) <= 0; };
        }
        return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_size(fr1, fr2) <= 0; };
    }

    int cmp_file_results_by_type(const FileResult& fr1, const FileResult& fr2) {
        if (fr1.file_type() == fr2.file_type()) {
            return cmp_file_results_by_path(fr1, fr2);
        }
        return fr1.file_type() < fr2.file_type() ? -1 : 1;
    }

    int cmp_file_results_by_type_ci(const FileResult& fr1, const FileResult& fr2) {
        if (fr1.file_type() == fr2.file_type()) {
            return cmp_file_results_by_path_ci(fr1, fr2);
        }
        return fr1.file_type() < fr2.file_type() ? -1 : 1;
    }

    std::function<bool(FileResult&, FileResult&)> FileResultSorter::get_cmp_file_results_by_type() const {
        if (m_settings.sort_descending()) {
            if (m_settings.sort_case_insensitive()) {
                return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_type_ci(fr2, fr1) <= 0; };
            }
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_type(fr2, fr1) <= 0; };
        }
        if (m_settings.sort_case_insensitive()) {
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_type_ci(fr1, fr2) <= 0; };
        }
        return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_type(fr1, fr2) <= 0; };
    }

    int cmp_file_results_by_lastmod(const FileResult& fr1, const FileResult& fr2) {
        if (fr1.last_mod() == fr2.last_mod()) {
            return cmp_file_results_by_path(fr1, fr2);
        }
        return fr1.last_mod() < fr2.last_mod() ? -1 : 1;
    }

    int cmp_file_results_by_lastmod_ci(const FileResult& fr1, const FileResult& fr2) {
        if (fr1.last_mod() == fr2.last_mod()) {
            return cmp_file_results_by_path_ci(fr1, fr2);
        }
        return fr1.last_mod() < fr2.last_mod() ? -1 : 1;
    }

    std::function<bool(FileResult&, FileResult&)> FileResultSorter::get_cmp_file_results_by_lastmod() const {
        if (m_settings.sort_descending()) {
            if (m_settings.sort_case_insensitive()) {
                return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_lastmod_ci(fr2, fr1) <= 0; };
            }
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_lastmod(fr2, fr1) <= 0; };
        }
        if (m_settings.sort_case_insensitive()) {
            return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_lastmod_ci(fr1, fr2) <= 0; };
        }
        return [](const FileResult& fr1, const FileResult& fr2) { return cmp_file_results_by_lastmod(fr1, fr2) <= 0; };
    }

    std::function<bool(FileResult&, FileResult&)> FileResultSorter::get_file_result_comparator() const {
        switch (m_settings.sort_by()) {
            case SortBy::FILENAME:
                return get_cmp_file_results_by_name();
            case SortBy::FILEPATH:
                return get_cmp_file_results_by_path();
            case SortBy::FILESIZE:
                return get_cmp_file_results_by_size();
            case SortBy::FILETYPE:
                return get_cmp_file_results_by_type();
            case SortBy::LASTMOD:
                return get_cmp_file_results_by_lastmod();
        }
        return get_cmp_file_results_by_path();
    }

    void FileResultSorter::sort(std::vector<FileResult>& file_results) const {
        const auto sort_comparator = get_file_result_comparator();
        std::ranges::sort(file_results, sort_comparator);
    }
}
